### PR TITLE
[AIRFLOW-5335] Simplify GCSHook test

### PR DESCRIPTION
**Pull request by kaxil**
SImplifying this test further so that it does not try to authenticate

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - Part of https://issues.apache.org/jira/browse/AIRFLOW-5335

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
SImplifying this test further so that  it does not try to authenticate

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: tests/contrib/hooks/test_gcs_hook.py [hunk 1]
```diff
@@ -71,10 +71,11 @@ def setUp(self):
             self.gcs_hook = gcs_hook.GoogleCloudStorageHook(
                 google_cloud_storage_conn_id='test')
 
-    def test_storage_client_creation(self):
+    @mock.patch(BASE_STRING.format("GoogleCloudBaseHook._get_credentials"))
+    def test_storage_client_creation(self, get_con_mock):
```
Comment (mik-laj) [inline tests/contrib/hooks/test_gcs_hook.py]:
```suggestion
    def test_storage_client_creation(self, mock_get_credentials):
```

The mock_ * pattern is added to the exclusion list in the pylint. It is never marked as an unused variable. Changing the variable name also makes it easier to understand what is in the variable.

File: tests/contrib/hooks/test_gcs_hook.py [hunk 2]
```diff
@@ -71,10 +71,11 @@ def setUp(self):
             self.gcs_hook = gcs_hook.GoogleCloudStorageHook(
                 google_cloud_storage_conn_id='test')
 
-    def test_storage_client_creation(self):
+    @mock.patch(BASE_STRING.format("GoogleCloudBaseHook._get_credentials"))
```
Comment (mik-laj) [inline tests/contrib/hooks/test_gcs_hook.py]:
```suggestion
    @mock.patch(BASE_STRING.format("GoogleCloudBaseHook._get_credentials"), return_value = mock.ANY)
```
It seems to me that makes the code easier to read.

File: tests/contrib/hooks/test_gcs_hook.py [hunk 3]
```diff
@@ -71,10 +71,11 @@ def setUp(self):
...
+    def test_storage_client_creation(self, get_con_mock):
+        get_con_mock.return_value = mock.ANY
```
Comment (mik-laj) [inline tests/contrib/hooks/test_gcs_hook.py]:
```suggestion
```

File: tests/contrib/hooks/test_gcs_hook.py (hunk 2 above)
Comment (mik-laj) [reply tests/contrib/hooks/test_gcs_hook.py]: But I think that it is worth to return some value here.
```suggestion
    @mock.patch(BASE_STRING.format("GoogleCloudBaseHook._get_credentials"), return_value = "CREDENTIALS")
```
`"CREDENTIALS"` is a marker

File: tests/contrib/hooks/test_gcs_hook.py [hunk 4]
```diff
@@ -71,10 +71,11 @@ def setUp(self):
...
         with mock.patch('google.cloud.storage.Client') as mock_client:
-            gcs_hook_1 = gcs_hook.GoogleCloudStorageHook()
-            gcs_hook_1.get_conn()
+            gcs_hook.GoogleCloudStorageHook().get_conn()
 
             # test that Storage Client is called with required arguments
             mock_client.assert_called_once_with(
```
Comment (mik-laj) [inline tests/contrib/hooks/test_gcs_hook.py]:
```suggestion
            mock_client.assert_called_once_with(
            credentials ="CREDENTIALS"
```
Checking if the marker has been passed.

**Comment by mik-laj** [comment]
```python
    @mock.patch(
    	BASE_STRING.format("GoogleCloudBaseHook._get_credentials_and_project_id"), 
    	return_value=("CREDENTIALS", "PROJECT_ID")
    )
    @mock.patch('google.cloud.storage.Client')
    def test_storage_client_creation(self, mock_client, mock_get_credentials_and_project_id):
        hook = gcs_hook.GoogleCloudStorageHook()
        result = hook.get_conn()
        # test that Storage Client is called with required arguments
        mock_client.assert_called_once_with(
            client_info=hook.client_info,
            credentials="CREDENTIALS",
            project="PROJECT_ID")
        self.assertEqual(mock_client.return_value, result)
```
I looked deeper and I have more comments about this code. I think we should mock another method because it calls the external library. I think it's worth adding assertions for the result. The code is clearer when only one way of mocking is used - the decorator. Using context manager and decorator in one test is pointless.

**Comment by kaxil** [comment]
I like the idea. Only 1 issue regarding client_info, it creates a separate object:

```
AssertionError: Expected call: Client(client_info=<google.api_core.gapic_v1.client_info.ClientInfo object at 0x110f774a8>, credentials='CREDENTIALS', project='PROJECT_ID')
Actual call: Client(client_info=<google.api_core.gapic_v1.client_info.ClientInfo object at 0x11f47a080>, credentials='CREDENTIALS', project='PROJECT_ID')
```

**Comment by mik-laj** [comment]
This should fix these errors.
```
    @mock.patch(
        'airflow.contrib.hooks.gcp_api_base_hook.GoogleCloudBaseHook.client_info',
        new_callable=PropertyMock,
        return_value="CLIENT_INFO"
    )
```

**Comment by kaxil** [comment]
@mik-laj Updated

**Review by mik-laj — APPROVED** [review]

**Comment by mik-laj** [comment]
Pylint is sad. I will fix it.

**Comment by kaxil** [comment]
Thanks @mik-laj

**Pull request merged**